pkg/scanner.go

```diff
@@ -41,26 +72,80 @@ func (s *Scanner) Resources() ([]resource.Resource, error) {
 				logrus.WithFields(logrus.Fields{
 					"id":   resource.TerraformId(),
 					"type": resource.TerraformType(),
-				}).Debug("Found cloud resource")
+				}).Debug("[DEPRECATED] Found cloud resource")
 			}
 			return res, nil
 		})
 	}
 
-	results := make([]resource.Resource, 0)
-loop:
-	for {
-		select {
-		case resources, ok := <-s.runner.Read():
-			if !ok || resources == nil {
-				break loop
+	return s.retrieveRunnerResults()
+}
+
+func (s *Scanner) scan() ([]resource.Resource, error) {
+	for _, enumerator := range s.remoteLibrary.Enumerators() {
+		enumerator := enumerator
+		s.runner.Run(func() (interface{}, error) {
+			resources, err := enumerator.Enumerate()
+			if err != nil {
+				return nil, err
 			}
-			results = append(results, resources.([]resource.Resource)...)
-		case <-s.runner.DoneChan():
-			break loop
-		}
+			for _, resource := range resources {
+				if resource == nil {
+					continue
+				}
+				logrus.WithFields(logrus.Fields{
+					"id":   resource.TerraformId(),
+					"type": resource.TerraformType(),
+				}).Debug("Found cloud resource")
+			}
+			return resources, nil
+		})
 	}
-	return results, s.runner.Err()
+
+	enumerationResult, err := s.retrieveRunnerResults()
+	if err != nil {
+		return nil, err
+	}
+
+	s.runner = parallel.NewParallelRunner(context.TODO(), 10)
+
+	for _, res := range enumerationResult {
+		res := res
+		s.runner.Run(func() (interface{}, error) {
+			fetcher := s.remoteLibrary.GetDetailFetcher(resource.ResourceType(res.TerraformType()))
+			if fetcher != nil {
+				// If we are in deep mode, retrieve resource details
+				if s.options.Deep {
+					resourceWithDetails, err := fetcher.ReadDetails(res)
+					if err != nil {
+						return nil, err
+					}
+					return []resource.Resource{resourceWithDetails}, nil
+				}
+			}
+			return []resource.Resource{res}, nil
+		})
+	}
+
+	return s.retrieveRunnerResults()
+}
+
+func (s *Scanner) Resources() ([]resource.Resource, error) {
+
+	resources, err := s.legacyScan()
+	if err != nil {
+		return nil, err
+	}
+
+	s.runner = parallel.NewParallelRunner(context.TODO(), 10)
```

Comment:
Why do we create twice the runner (LOC 110 and 140) ?

Reply:
We should recreate the runner after reading results from it, maybe it's better to not reuse the original runner and have another runner. I don't know if they were designed to be used multiples times sequentially ? Maybe we should create two runner, one for enumeration and one for details fetching ?

Reply:
Maybe @moadibfr can help on this one since he's the original creator of the runner.

Reply:
I don't think they could be used two time we didn't test for this. But it might be slightly more readable to have two of them ?